component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -209,9 +257,155 @@ protected boolean authenticate(X509Certificate registeredCert, X509Certificate r
         return trustedCert;
     }
 
+    /**
+     * Authenticate the client by comparing the attributes retrieved from the JWKS endpoint of the registered public
+     * certificate against the public key of the certificate presented at TLS hand shake for authentication.
+     *
+     * @param jwksUri     JWKS URI registered at service provider configuration.
+     * @param requestCert X.509 certificate presented to server during TLS hand shake.
+     * @return Whether the client was successfully authenticated or not.
+     */
+    private boolean authenticate(URL jwksUri, X509Certificate requestCert) throws OAuthClientAuthnException {
+
+        try {
+            return isAuthenticated(getResourceContent(jwksUri), requestCert);
+
+        } catch (IOException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR,
+                    "Error occurred while opening HTTP connection for the JWKS URL : " + jwksUri, e);
+        } catch (NoSuchAlgorithmException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR,
+                    "Error occurred while generating thumbprint for registered certificate ", e);
+        } catch (CertificateException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR,
+                    "Error occurred while parsing certificate retrieved from JWKS endpoint ", e);
+        }
+    }
+
+    /**
+     * Authenticate the client by iterating through the JSON Array and matching each attribute.
+     *
+     * @param resourceArray Json Array retrieved from JWKS endpoint
+     * @param requestCert   X.509 certificate presented to server during TLS hand shake.
+     * @return Whether the client was successfully authenticated or not.
+     */
+    private boolean isAuthenticated(JsonArray resourceArray, X509Certificate requestCert)
+            throws CertificateException, OAuthClientAuthnException, NoSuchAlgorithmException {
+
+        for (JsonElement jsonElement : resourceArray) {
+            JsonElement attributeValue = jsonElement.getAsJsonObject().get(X5T);
+            if (attributeValue != null) {
+                if (attributeValue.getAsString().equals(MutualTLSUtil.getThumbPrint(requestCert))) {
```

Comment:
improve this logic by merging these two if blocks to one

Reply:
Fixed https://github.com/wso2-extensions/identity-oauth-addons/pull/52